blog/2019-09-ngrx-data-views/README.md

```diff
@@ -0,0 +1,419 @@
+
+If you are however used to have business objects around (enterprise developer?) and you want to have at least some kind of logic on your objects, data views can be a perfect place to implement it. As those objects basically spawn off the selectors, just outside of the store, you can add methods to them or even elevate simple properties to unserializable objects here.
+
+Let's consider the following scenario: You want to keep the publishing date on the `Book` objects and you also want to have some accessor there that tells you if the book is new because it was published this year. In the book model in the NgRx store you would have to keep the published date as a `number` because instances of the `Date` class are not serializable. Concerning the "new" property, you don't want to put it into the store at all, as it is derived from the publishing date of the book and keeping it in the store would be redundant and therefore contradicting normalization. Also, you also cannot add it as a method to the model because functions are not serializable either.
+
+With data views you would just solve all this in the `BookView` interface and the corresponding selector:
+
+```ts
+export interface Book {
+  ...
+  published: number;
+}
+
+export interface BookView {
+  ...
+  published: Date;
+  isNew(): boolean;
+}
+
+export const calculateNew = (book: Book) => {
+  return new Date().getFullYear() === new Date(book.published).getFullYear();
+};
+
+export const getBookView = () =>
+  createSelector(
+    ...
+    (book, authors, tags): BookView =>
+      book && {
+        ...
+        published: new Date(book.published),
+        isNew: () => calculateNew(book),
+      }
+  );
+```
+
+The result can be observed in the next [StackBlitz Example](https://stackblitz.com/github/dhhyi/ngrx-data-views/tree/data-views-as-business-objects?file=src%2Fapp%2Fstore%2Fbook-view%2Findex.ts&devtoolsheight=50). As you can see in the console, solving it this way leads to many calls to the "new" method provided by the data view. The reason lies within the Angular change detection.
+
+## Memoized Data View Methods
+
+When developing, you are composing all components living in an Angular application into a tree for rendering. Likewise Angular keeps a tree of corresponding change detectors for optimizing view updates. All bindings in templates referencing component properties are checked in every change detector cycle. But when one is using call expressions on the template (which we would be doing when accessing `book.isNew()`), the binding cannot be properly cached and thus has to be re-evaluated in every cycle.
```

Comment:
This is based on suspicion and should be researched and or backed up by source code references :sweat_smile:

Reply:
Each expression in the template is being checked and re-evaluated in every CD cycle, no matter whether it's a property or a method/function.
There is no cache, but the CD just compares old and new value and updates the view if they differ. This is not problematic for simple values, but when functions and calculations are involved, this is a problem and should be avoided.

Reply:
As i thought :sweat_smile:

Reply:
made less assumptions